src/mbgl/tile/geometry_tile_worker.cpp

```diff
@@ -143,24 +154,46 @@ void GeometryTileWorker::setPlacementConfig(PlacementConfig placementConfig_, ui
     }
 }
 
+void GeometryTileWorker::onGlyphsAvailable(GlyphPositionMap glyphs) {
+    glyphPositions = std::move(glyphs);
+    waitingForGlyphs = false;
+    symbolDependenciesChanged();
+}
+
+void GeometryTileWorker::onIconsAvailable(IconAtlasMap icons_) {
+    icons = std::move(icons_);
+    waitingForIcons = false;
+    symbolDependenciesChanged();
+}
+
+bool GeometryTileWorker::hasPendingSymbolDependencies() const {
+    return waitingForGlyphs || waitingForIcons;
+}
+
 void GeometryTileWorker::symbolDependenciesChanged() {
     try {
         switch (state) {
-        case Idle:
-            if (hasPendingSymbolDependencies()) {
+        case NeedPlacement:
+            if (!hasPendingSymbolDependencies()) {
```

Comment:
Those are good suggestions. Making the changes caused some problems with a part of the logic I hadn't fully worked through before (that is: what happens when messages arrive but the placement config or data isn't set yet). I addressed those and everything looked great aside from the state transition diagram getting more complicated... but now I'm seeing an intermittent memory corruption problem. So that's reassuring.

I'll take another stab at this tomorrow.


![geometrytileworker2](https://cloud.githubusercontent.com/assets/375121/24384465/db6e7aa4-1318-11e7-833e-6d800c7189e7.png)